### ISSUE TITLE
About  normalization of Text Embeddings

### ISSUE THREAD

**Issue by WangdongZhao99** [FereshteShakeri/Histo-TransCLIP#3]
Thank for ur work! 
I found multiple templates in the code. prepare_objects in TransCLIP_solver/TransCLIP_utilits-py(line 272).
`

    if len(clip_prototypes.shape) == 3:  # use more than 1 template
        clip_prototypes = clip_prototypes.mean(dim=0) # clip_prototypes[0] use only the first one clip_prototypes.mean(dim=0)

    if few_shot:
        support_features = support_features.squeeze().float()
        val_features = val_features.cuda().float()
        val_labels = val_labels.cuda()
        # find the nearest neighbor of each validation sample in the query set for prediction on the validation set:
        neighbor_index = torch.argmax(val_features @ query_features.T, dim=1)

    clip_logits = 100 * query_features @ clip_prototypes
`
visual embedding and text embedding are normalized before this, but here the text embedding (num_templates * dim * C) is not normalized after being mean (dim * C). Instead, directly calculate cosine similarity with visual embedding below. Although this approach has generally improved the performance compared to normalization, it is not rigorous enough in theory. In the original TransCLIP, the code used the first text prototype instead of the averaged one. I would like to know why normalization is not performed and look forward to your reply.
Here are the results I reproduced and the results after normalizing the mean.

![Image](https://github.com/user-attachments/assets/69d645de-eee7-4bf4-a043-e041c232be22)

**Comment by WangdongZhao99** [comment]
`class NCT(DatasetBase):

    dataset_dir = "NCT-CRC-HE-100K"

    def __init__(self, root, preprocess):
        
        self.dataset_dir = os.path.join(root, self.dataset_dir)
        self.image_dir = os.path.join(self.dataset_dir, "NCT-CRC-HE-100K")
        self.image_dir_test = os.path.join(self.dataset_dir, "CRC-VAL-HE-7K")

        text_file = os.path.join(self.dataset_dir, "classnames.txt")
        # classnames, labels = self.read_classnames(text_file)
        self.template = templates
        train_path = os.path.join(self.dataset_dir, "NCT-CRC-HE-100K")
        test_path = os.path.join(self.dataset_dir, "CRC-VAL-HE-7K")

        train, val = self.read_data(train_path, "train")
        test,_ = self.read_data(test_path, "test")

        super().__init__(train_x=train, val=val, test=test)

    def read_data(self, data_path, split):
        
        if split == "test":
            image_dir = self.image_dir
        else:
            image_dir = self.image_dir
        folders = listdir_nohidden(image_dir, sort=True)
        folders = [f for f in folders if f not in TO_BE_IGNORED]
        items = []
`
In addition, I found an issue in datasets/nct.py. In lines 59 and 60, you defined the image paths for train and test. In line 69, you passed in the path for the test image, but in line 76, you called the path for train. Is this really the training set used here?

**Comment by FereshteShakeri** [comment]
Hello Wang,

Thanks for your interest in our work. you're right that after averaging the normalized text embeddings, the result isn’t necessarily unit-length, so technically it should be normalized again before computing cosine similarity. I skipped that step which seems it gave slightly better results in practice, even if it’s not fully rigorous. That said, I agree it’s worth re-evaluating.

Also, thanks for spotting the issue in datasets/nct.py. The original version was actually using the test set — I just hadn’t updated the code. That’s now fixed. Appreciate you pointing it out — we had missed putting the latest update.

**Comment by WangdongZhao99** [comment]
Thanks for your reply
